academy/ibc-dev/ibcmw-integrate.md

```diff
@@ -0,0 +1,156 @@
+# Integrating IBC middleware into a chain
+
+<HighlightBox type="learning">
+
+Learn how to integrate IBC middleware(s) with a base application to your chain. The following document only applies for Cosmos SDK chains.
+
+</HighlightBox>
+
+We've now seen how to develop an IBC middleware, the final step in order to use it is to integrate it into the chain. We can distinguish between a single piece of middleware or a stack of middleware and between middleware that is stateful or stateless.
+
+To integrate our middleware (stack) we must do the following in `app.go`:
+
+- If the middleware is maintaining its own state and/or processing SDK messages, then it should create and register its SDK module **only once** with the module manager
+
+    ```go
+    //app.go
+
+    // create a keeper for the stateful middleware
+    mw1Keeper := mw1.NewKeeper(storeKey1)
+    mw3Keeper := mw3.NewKeeper(storeKey3)
+
+    // instantiate the middleware as IBCModules
+    mw1IBCModule := mw1.NewIBCModule(mw1Keeper)
+    mw2IBCModule := mw2.NewIBCModule() // middleware2 is stateless middleware
+    mw3IBCModule := mw3.NewIBCModule(mw3Keeper)
```

Comment:
@AdityaSripal do we have to go through this step to instantiate as IBCModule first?

If I have a single middleware not used in a stack, say `mw4`, could I just do the following:

```
mw4Keeper := mw4.NewKeeper(storeKey4)
mw4IBCMiddleware := mw4.NewIBCMiddleware(customIBCModule1, mw4Keeper)
```

or should I still rather do:
```
mw4Keeper := mw4.NewKeeper(storeKey4)
mw4IBCModule := mw4.NewIBCModule(mw4Keeper)
mw4IBCMiddleware := mw4.NewIBCMiddleware(customIBCModule1, mw4Keeper)
```

Reply:
Thanks for pointing this out Thomas.

`NewIBCMiddleware` is a function defined by each custom middleware, so if it is defined on the middleware you want to use, you can use it to create the final stack that you wish.

Whether or not you need an `AppModule` defined for your middleware really depends on the middleware in question. Defining the module in `app.go` is necessary if you then need to separately wire the `middlewareModule` into app.go's module wiring functions (e.g. InitGenesis/BeginBlock/MsgRouter/etc)

So as a short answer, I would say if you have a **stateful** middleware (or middleware that will execute independent messages) then you will most likely need to define the `middleware` module so that you can wire in all the state tracking/execution logic into app.go.

If you have **stateless** middleware (e.g. a simple filtering mw), then you will not need to define the module, you can simply wrap the underlying application with your middleware